### ISSUE TITLE
lab2

### ISSUE THREAD

**Issue by dlcor** [dlcor/AI#2]
# Импорт необходимых библиотек
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from matplotlib.colors import ListedColormap

# Импорт компонентов из scikit-learn
from sklearn.preprocessing import LabelEncoder, StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import GaussianNB
from sklearn.svm import SVC
from sklearn.tree import DecisionTreeClassifier, plot_tree
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score

# Настройка бэкенда для matplotlib
import matplotlib

matplotlib.use('Qt5Agg')  # Используем Qt5 для отрисовки графиков
import matplotlib.pyplot as plt

# === 1. Загрузка данных ===
df = pd.read_csv("BIKE DETAILS.csv")

# === 2. Предобработка данных ===
# Заполнение пропусков в цене медианным значением
df["ex_showroom_price"] = df["ex_showroom_price"].fillna(df["ex_showroom_price"].median())

# Создание бинарной целевой переменной (мотоциклы после 2015 года)
df["Target"] = (df["year"] >= 2015).astype(int)

# Удаление ненужных столбцов
df.drop(columns=["year", "name"], inplace=True)

# Преобразование категориальных переменных в числовые
df_encoded = pd.get_dummies(df, columns=["seller_type", "owner"], drop_first=True)

# Разделение на матрицу признаков и целевую переменную
X = df_encoded.drop(columns=["Target"])
y = df_encoded["Target"]

# === 3. Разделение данных ===
# Стратифицированное разделение на train/test (сохраняем распределение классов)
X_train, X_test, y_train, y_test = train_test_split(
    X, y,
    test_size=0.2,
    random_state=42,
    stratify=y
)

# === 4. Масштабирование признаков ===
# Нормализация данных для алгоритмов, чувствительных к масштабу
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

# === 5. Инициализация моделей ===
models = {
    "Naive Bayes": GaussianNB(),  # Вероятностный классификатор
    "SVM": SVC(kernel="rbf", probability=True, max_iter=20000),  # Метод опорных векторов
    "Decision Tree": DecisionTreeClassifier(),  # Дерево решений
    "Rule-Based Classifier": DecisionTreeClassifier(max_depth=3),  # Упрощенное дерево (правила)
    "Random Forest": RandomForestClassifier(n_estimators=100)  # Ансамбль деревьев
}

# Функция для обучения и оценки модели
def evaluate_model(model, X_train, y_train, X_test, y_test):
    model.fit(X_train, y_train)
    y_pred = model.predict(X_test)
    return (
        classification_report(y_test, y_pred, output_dict=True),
        confusion_matrix(y_test, y_pred),
        accuracy_score(y_test, y_pred),
        y_pred
    )

# === 6. Визуализация результатов ===
def plot_confusion_matrix(y_test, y_pred, model_name):
    # Построение тепловой карты матрицы ошибок
    cm = confusion_matrix(y_test, y_pred)
    plt.figure(figsize=(5, 4), num=f"Confusion Matrix - {model_name}")
    sns.heatmap(cm, annot=True, fmt="d", cmap="Blues",
                xticklabels=["До 2015", "После 2015"],
                yticklabels=["До 2015", "После 2015"])
    plt.xlabel("Предсказано")
    plt.ylabel("Фактическое")
    plt.title(f"Матрица ошибок - {model_name}")
    plt.show()

def plot_decision_tree(model, feature_names, model_name):
    # Визуализация структуры дерева решений
    plt.figure(figsize=(15, 8), num=f"Decision Tree - {model_name}")
    plot_tree(model, feature_names=feature_names,
              filled=True, rounded=True,
              class_names=["До 2015", "После 2015"])
    plt.title(f"Дерево решений - {model_name}")
    plt.show()

# === 7. Обучение и оценка моделей ===
results = {}
for name, model in models.items():
    print(f"\nОбучение модели: {name}")
    report, cm, acc, y_pred = evaluate_model(model, X_train, y_train, X_test, y_test)

    # Сохранение результатов
    results[name] = {
        "classification_report": report,
        "confusion_matrix": cm,
        "accuracy": acc
    }

    # Вывод метрик
    print(f"Accuracy: {acc:.4f}")
    print(classification_report(y_test, y_pred))

    # Визуализация
    plot_confusion_matrix(y_test, y_pred, name)
    if name in ["Decision Tree", "Rule-Based Classifier"]:
        plot_decision_tree(model, X.columns, name)

# === 8. Экспорт результатов ===
# Формирование сводной таблицы метрик
results_summary = []
for name, metrics in results.items():
    results_summary.append({
        "Model": name,
        "Accuracy": metrics["accuracy"],
        "Precision (avg)": metrics["classification_report"]["weighted avg"]["precision"],
        "Recall (avg)": metrics["classification_report"]["weighted avg"]["recall"],
        "F1-Score (avg)": metrics["classification_report"]["weighted avg"]["f1-score"],
    })

results_df = pd.DataFrame(results_summary)
results_df.to_csv("results_summary.csv", index=False)

# === 9. 2D визуализация границ решений ===
def plot_decision_boundary(model, X, y, title):
    # Визуализация разделяющей поверхности для двух признаков
    x_min, x_max = X.values[:, 0].min() - 1, X.values[:, 0].max() + 1
    y_min, y_max = X.values[:, 1].min() - 1, X.values[:, 1].max() + 1
    xx, yy = np.meshgrid(np.arange(x_min, x_max, 0.05),
                         np.arange(y_min, y_max, 0.05))

    # Предсказание для сетки точек
    Z = model.predict(pd.DataFrame(np.c_[xx.ravel(), yy.ravel()],
                                   columns=X.columns)).reshape(xx.shape)

    # Отрисовка границ
    plt.figure(figsize=(10, 6), num=title)
    plt.contourf(xx, yy, Z, alpha=0.75, cmap=ListedColormap(('red', 'green')))
    plt.scatter(X.values[:, 0], X.values[:, 1], c=y, edgecolors='k', alpha=0.8)
    plt.title(title)
    plt.xlabel("ex_showroom_price")
    plt.ylabel("km_driven")
    plt.show()

# Обучение и визуализация для двух признаков
feature1, feature2 = "ex_showroom_price", "km_driven"
X_train_2d = pd.DataFrame(X_train, columns=X.columns)[[feature1, feature2]]
X_test_2d = pd.DataFrame(X_test, columns=X.columns)[[feature1, feature2]]

models_2d = {
    "Naive Bayes (2D)": GaussianNB(),
    "SVM (2D)": SVC(kernel="rbf", probability=True, max_iter=20000),
    "Decision Tree (2D)": DecisionTreeClassifier(),
    "Rule-Based (2D)": DecisionTreeClassifier(max_depth=3),
    "Random Forest (2D)": RandomForestClassifier(n_estimators=100)
}

for name, model in models_2d.items():
    model.fit(X_train_2d, y_train)
    plot_decision_boundary(model, X_train_2d, y_train, f"{name} (Train Set)")
    plot_decision_boundary(model, X_test_2d, y_test, f"{name} (Test Set)")
